File: frontend/de/unika/ipd/grgen/ir/stmt/ConditionStatement.java
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 5.0
 * Copyright (C) 2003-2020 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3 (see LICENSE.txt included in the packaging of this file)
 * www.grgen.net
 */

/**
 * @author Edgar Jakumeit
 */

package de.unika.ipd.grgen.ir.stmt;

import java.util.Collection;
import java.util.LinkedList;

import de.unika.ipd.grgen.ir.NeededEntities;
import de.unika.ipd.grgen.ir.expr.Expression;

/**
 * Represents a condition statement in the IR.
 */
public class ConditionStatement extends NestingStatement
{
	private Expression conditionExpr;
	private Collection<EvalStatement> falseCaseStatements = null;

	public ConditionStatement(Expression conditionExpr)
	{
		super("condition statement");
		this.conditionExpr = conditionExpr;
	}

	public void addFalseCaseStatement(EvalStatement falseCaseStatement)
	{
		if(falseCaseStatements == null)
			falseCaseStatements = new LinkedList<EvalStatement>();
		falseCaseStatements.add(falseCaseStatement);
	}

	public Expression getConditionExpr()
	{
		return conditionExpr;
	}

	public Collection<EvalStatement> getFalseCaseStatements()
	{
		return falseCaseStatements;
	}

	@Override
	public void collectNeededEntities(NeededEntities needs)
	{
		conditionExpr.collectNeededEntities(needs);
		for(EvalStatement trueCaseStatement : statements) {
			trueCaseStatement.collectNeededEntities(needs);
		}
		if(falseCaseStatements != null) {
			for(EvalStatement falseCaseStatement : falseCaseStatements) {
				falseCaseStatement.collectNeededEntities(needs);
			}
		}
	}
}

File: frontend/de/unika/ipd/grgen/ir/stmt/ConditionStatement.cs
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 8.1
 * Copyright (C) 2003-2026 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3, some components/parts use different licenses (see LICENSE.txt included in the packaging of this file)
 * www.grgen.de / www.grgen.net
 */

/// <summary>
/// @author Edgar Jakumeit
/// </summary>

namespace de.unika.ipd.grgen.ir.stmt
{

using System.Collections.Generic;

using NeededEntities = de.unika.ipd.grgen.ir.NeededEntities;
using Expression = de.unika.ipd.grgen.ir.expr.Expression;

/// <summary>
/// Represents a condition statement in the IR.
/// </summary>
public class ConditionStatement : BlockNestingStatement
{
	private Expression conditionExpr;
	private List<EvalStatement> falseCaseStatements = null;

	public ConditionStatement(Expression conditionExpr)
		: base("condition statement")
	{
		this.conditionExpr = conditionExpr;
	}

	public virtual void AddFalseCaseStatement(EvalStatement falseCaseStatement)
	{
		if(falseCaseStatements == null)
			falseCaseStatements = new List<EvalStatement>();
		falseCaseStatements.Add(falseCaseStatement);
	}

	public virtual Expression ConditionExpr
	{
		get
		{
		return conditionExpr;
		}
	}

	public virtual ICollection<EvalStatement> FalseCaseStatements
	{
		get
		{
		return falseCaseStatements != null ? falseCaseStatements.AsReadOnly() : null;
		}
	}

	public override void CollectNeededEntities(NeededEntities needs)
	{
		conditionExpr.CollectNeededEntities(needs);
		foreach(EvalStatement trueCaseStatement in statements)
			trueCaseStatement.CollectNeededEntities(needs);
		if(falseCaseStatements != null)
		{
			foreach(EvalStatement falseCaseStatement in falseCaseStatements)
				falseCaseStatement.CollectNeededEntities(needs);
		}
	}
}

}
